sdb.3
.TH SDB 3 LOCAL

.SH NAME

sdb_init, sdb_open, sdb_query, sdb_close \- the Simple Database Library.

.SH SYNOPSIS

.B #include <sdb.h>
.sp
.BI "void sdb_init();"
.br
.BI "char *sdb_open(char *" url ");"
.br
.BI "void sdb_close(char *" id ");"
.br
.BI "int sdb_query(char *" url ", char *" query ", int (*" callback ")(int, char **, void *), void *" closure ");"

.SH DESCRIPTION

The SDB library allows applications to support multiple database management
systems with negligeable overhead, in terms of code as well as system
resources.

.PP
.B sdb_init()
initializes the library and registers the database drivers. It is
not necessary to call sdb_init explicitly, since it will be done
automatically when needed.
.PP
.B sdb_open()
opens a database connection that can be used for multiple queries. This
is optional; calling sdb_query directly will simply open and close the
connection for each query. sdb_open returns a connection id which is used
in place of the url in calls to sdb_query.
.PP
.B sdb_query()
calls the callback once for each row returned. No rows does not indicate
an error condition. sdb_query returns the number of rows or -1 for error.
The callback takes three arguments, an integer indicating the number of
columns in the result, an array of pointers to the fields and a pointer
to some arbitrary data that the callback might need. Values are always
returned as strings.
.PP
.B sdb_close()
closes the database connection opened by sdb_open.

.SH EXAMPLES

This minimal program runs queries from the command line.

.nf

#include <stdio.h>
#include <stdlib.h>
#include <sdb.h>

static int callback(int n, char **p, void *closure)
{
    int i;
    for (i = 0; i < n; i++) {
        printf("%s\\t", p[i]);
    }
    printf("\n");
    return 0;
}

int main(int argc, char **argv)
{
    if (argc != 3) {
        fprintf(stderr, "Usage: sdb_demo url query\\n");
        return EXIT_FAILURE;
    }
    sdb_query(argv[1], argv[2], callback, NULL);
    return EXIT_SUCCESS;
}

.fi

This program can be used to authenticate Squid proxy users.

.nf

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sdb.h>

static int cb_db(int n, char **p, void *closure)
{
    return 0;
}

int main(int argc, char **argv)
{
    char *url, query[1024];
    int n;
    char buf[256];
    char *user, *passwd, *p;
    setbuf(stdout, NULL);
    if (argc != 2) {
        fprintf(stderr, "Usage: sdb_auth url\\n");
        exit(1);
    }
    url = argv[1];
    while (fgets(buf, 256, stdin) != NULL) {
        if ((p = strchr(buf, '\\n')) != NULL)
            *p = '\\0';          /* strip \\n */
        if ((user = strtok(buf, " ")) == NULL) {
            printf("ERR\\n");
            continue;
        }
        if ((passwd = strtok(NULL, "")) == NULL) {
            printf("ERR\\n");
            continue;
        }
        sprintf(query,
                "select * from htpasswd "
                "where user = '%s' "
                "and passwd = '%s'",
                user, passwd);
        n = sdb_query(url, query, cb_db, NULL);
        if (n < 1) {
            printf("ERR\\n");
        } else {
            printf("OK\\n");
        }
    }
    exit(0);
}

.fi

.SH SEE ALSO

sdbd.8

Example clients in sdb_client.c and sdbd_client.c.

.SH AUTHOR

Copyright (c) 2000-2005  Ulric Eriksson <ulric@siag.nu>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public
License as published by the Free Software Foundation; either
version 2 of the Licence, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston,
MA 02111-1307, USA.



ChangeLog

150222	Updated config.in.

101005	More tweaking of the configuration script for compatibility
	with the latest instant oracle client.
	Released 0.10.1.

080430	Oracle driver again: increased the size of the hda to 512 bytes
	for compatibility with 64-bit platforms.
	Released 0.10.0.

080427	Made the Oracle driver build compatible with Oracle Instant Client.

080425	Added db option to oracle url. These are now equivalent:
	oracle:uid=username/password@database
	oracle:uid=username:pwd=password:db=database

080311	Added --disable-shared and --disable-static options to
	the configure script.
	Released 0.9.0.

080307	In sdb.c, function sdb_close: don't call the driver's close
	function if the db is null (i.e. there is nothing to close).

080306	Started integrating Freetds for Microsoft SQL and Sybase.
	Queries are submitted but returned values are not handed
	off to the callback.

080305	Oracle driver updated for 11g.
	Make sure that the Oracle driver treats null values as NULL.
	Released 0.8.0.

080304	Initial support for Ingres.

080229	Added support for shsql/quisp.
	Changed clilen type to socklen_t in sdbd.c.
	Released 0.7.0.

051029	Added -c option to sdb_client to read an sql query from the
	command line rather than interactively.
	Released 0.6.2.

050219	In sdb_query: if the url refers to an already opened connection,
	check that the connection is in fact opened. This is important
	for mistyped urls.
	Released 0.6.1.

050110	Added --platform flag to the configure script.

050110	Added support for Sqlite 3.
	Added another minimal example to the manpage.
	Released 0.6.0.

050109	Dumped automake. Its self-incompatibility drives me nuts!

040506	More bugfixes from Guillaume Cottenceau to the Postgres
	and Mysql drivers: don't segfault if sdb_close is called
	with a NULL argument.
	Released 0.5.5.

040423	Bugfix in the postgres driver: the closer should be called
	sdb_postgres_close. Released 0.5.4.

040421	Patch for persistent Postgres connections from
	Guillaume Cottenceau <gc at mnc.ch>. Released 0.5.3.

040227	Check for Solaris specific libraries.
	Added patch for persistent Oracle connections
	from Denis Ovsienko <pilot@altlinux.ru>.

020828	Do not link sdb_client against .libs/libsdb.so. It breaks the
	build on certain platforms.
	sdb_client: only print prompt if stdin is a terminal.
	sdb_client: Option -s changes the separator in output.
	Released 0.5.1.

020630	Added a driver by Lahcen EL HADDATI <elhaddati@hotmail.com>
	which accepts queries on plain text files in a SQL-like language.
	Released 0.5.0.

020621	Fixed bug in sdbd: null values in the result set caused the
	process to crash. Now it returns an empty string, which is wrong
	but better than crashing.
	Check memory allocations better in sdbd_driver.

020620	Mimer driver that actually works.
	Released 0.4.1.

020530	MiniSQL driver. Use --with-msql=/path/to/msql if it is not
	installed in /usr/local/Hughes.
	Released 0.4.0.

020526	Implemented open and close for mysql driver.
	Wrote manpages for sdb_open, sdb_close, sdb_query, sdb_init and sdbd.
	Released 0.3.0.

020522	Added sdb_open and sdb_close for connections that are to be
	used for more than one query.
	Implemented open and close for sqlite driver.
	Modified the sdbd protocol to allow connections to be used for
	multiple queries. Updated sdbd.c and sdbd_driver.c to use the new
	protocol if it is implemented on both server and client.

020520	Added a check in sdb_query to call sdb_init if it hasn't already
	been done. This makes sdb_init calls from applications optional.

020515	Improved error checking in gdbm_driver.c.

020514	Added squid_sdb_auth.c, a program to authenticate Squid proxy
	users to any database supported by libsdb.
	Released 0.2.4.

020503	Renamed all the drivers from XXX.c to XXX_driver.c.
	Wrote a new driver (sdbd_driver.c) for proxied queries.
	The new driver is designed so it can be compiled statically
	into an application without the rest of libsdb.

020429	Added optional argument to --with-odbc option, plus the necessary
	tweaking to get the odbc driver running on Windows.
	Changed #include <mysql/mysql.h> to just #include <mysql.h>.
	Released 0.2.3.

020425	New flags -d and -f to sdbd control debug message level and
	background/foreground operation.
	Made sdbd less eager to error out at every problem.

020209	Modification to postgres driver to let it work with
	user/password login (srosso@libero.it).
	Released 0.2.2.

020105	Fixed compiler warning in sdbd_client.c (2nd arg to connect).

011220	Updated automake, autoconf and libtool.

011204	Updated configuration for Mysql.

010910	SO_REUSEADDR in sdbd.c.

010611	Fixed silly bug in sdbd.c and sdbd_client.c: readdata and writedata
	would fail when data was read and written in small segments.
	Released 0.2.1.

010610	Sample client for sdbd in sdbd_client. Released 0.2.0.

010601	Proxy daemon called sdbd.

010520	Updated documentation and sdb_client.
	Released 0.1.0.

010517	Added trivial gdbm driver (adapted from Siag).

010515	Added autoconf and libtool.

010513	Released 0.0.1.

010512	Cleaned up namespace by prefixing all globals with sdb_.

010511	Started project by copying database code from Ucronadm.



sdb.c
/*
   Copyright (C) 2000-2008  Ulric Eriksson <ulric@siag.nu>
 
   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License as published by the Free Software Foundation; either
   version 2 of the Licence, or (at your option) any later version.

   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.

   You should have received a copy of the GNU Library General Public
   License along with this library; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston,
   MA 02111-1307, USA.
 */
 
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <sys/time.h>
#include <unistd.h>
#include "common.h"
#include "sdb.h"

#define CONN_MAX 100

extern void sdb_init_gdbm(void);	/* gdbm_driver.c */
extern void sdb_init_ingres(void);	/* ingres_driver.c */
extern void sdb_init_lago(void);        /* lago_driver.c */
extern void sdb_init_mimer(void);       /* mimer_driver.c */
extern void sdb_init_msql(void);	/* msql_driver.c */
extern void sdb_init_mysql(void);       /* mysql_driver.c */
extern void sdb_init_odbc(void);	/* odbc_driver.c */
extern void sdb_init_oracle(void);      /* oracle_driver.c */
extern void sdb_init_postgres(void);    /* postgres_driver.c */
extern void sdb_init_shsql(void);	/* shsql_driver.c */
extern void sdb_init_sqlite(void);      /* sqlite_driver.c */
extern void sdb_init_sqlite3(void);     /* sqlite3_driver.c */
extern void sdb_init_tds(void);		/* tds_driver.c */
extern void sdb_init_text(void);	/* text_driver.c */

static struct sdb_driver {
	char *name;
	int (*driver)(void *, char *, char *,
			int (*)(int, char **, void *), void *);
	void *(*open)(char *);
	int (*close)(void *);
} drivers[20];

static int ndrivers = 0;

static struct sdb_conn {
	char *url;
	char *index;
	int driver;
	void *db;
} conn[CONN_MAX];

char *sdb_url_value(char *p, char *k)
{
        int len = strlen(k);
        char b[1024];
        int i;
 
	if (sdb_debuglevel) sdb_debug("sdb_url_value(%s, %s)", p, k);

        while (p) {
                if (!sdb_strncasecmp(p, k, len) && p[len] == '=') {
                        p = p+len+1;
                        for (i = 0; p[i] && p[i] != ':'; i++)
                                b[i] = p[i];
                        b[i] = '\0';
			if (sdb_debuglevel) {
				sdb_debug("Key %s found, value '%s'", k, b);
			}
                        return sdb_strdup(b);
                }
                p = strchr(p, ':');
                if (p) p++;
        }
	if (sdb_debuglevel) sdb_debug("Key %s not found, returning NULL", k);
        return NULL;
}

void sdb_register_driver(char *name,
	int (*driver)(void *, char *, char *,
			int (*)(int, char **, void *), void *),
	void *(*open)(char *), int (*close)(void *))
{
	if (sdb_debuglevel) sdb_debug("sdb_register_driver(%s, %p, %p, %p)\n",
				name, driver, open, close);
	drivers[ndrivers].name = name;
	drivers[ndrivers].driver = driver;
	drivers[ndrivers].open = open;
	drivers[ndrivers].close = close;
	ndrivers++;
}

/* Return index of driver or -1 for failure */
static int lookup_driver(char *url)
{
	char driver[1024], *p;
	int n, i;

	if (url == NULL) return -1;
	if (sdb_debuglevel) sdb_debug("lookup_driver(%s)", url);
	p = strchr(url, ':');
	if (p) n = p-url;
	else n = strlen(url);
	if (n > 1000) n = 1000;
	strncpy(driver, url, n);
	driver[n] = '\0';
	if (sdb_debuglevel) sdb_debug("driver name = '%s'", driver);
	for (i = 0; drivers[i].name; i++) {
	if (sdb_debuglevel) sdb_debug("compare with '%s'", drivers[i].name);
		if (!sdb_strcasecmp(drivers[i].name, driver)) {
			return i;
		}
	}
	return -1;
}

/*
;@db_query(url, query)
;@Makes database queries. <a href="db.html">More information</a>.
;@db_query("odbc:postgresql", "select * from foo")
;@
*/
int sdb_query(char *url, char *query,
	int (*callback)(int, char **, void *), void *closure)
{
        char *u, *p;
        int i, j;
	void *db;
 
        if (url == NULL || query == NULL) return -1;
	sdb_init();
 
        u = sdb_strdup(url);
        p = strchr(u, ':');
	if (p == NULL) {
		j = atoi(u);
		if (j < 0 || j >= CONN_MAX || conn[j].url == NULL) {
			sdb_debug("%d: no such connection", j);
			return -1;
		}
		if (strcmp(u, conn[j].index)) {
			sdb_debug("Index %s does not match %s",
				u, conn[j].index);
			return -1;
		}
		i = conn[j].driver;
		db = conn[j].db;
		p = conn[j].url;
	} else {
		i = lookup_driver(u);
		if (i == -1) {
			sdb_debug("No driver for %s", u);
			return -1;
		}
		db = NULL;
        	*p++ = '\0';
	}
 
	return (*drivers[i].driver)(db, p, query, callback, closure);
}

char *sdb_open(char *url)
{
	char b[100];
	void *db;
	int i, j;

	sdb_init();

	i = lookup_driver(url);
	if (i == -1) {
		sdb_debug("No driver for %s", url);
		return NULL;
	}
	if (drivers[i].open == NULL) db = NULL;
	else db = (*drivers[i].open)(url+strlen(drivers[i].name)+1);
	for (j = 0; j < CONN_MAX; j++) {
		if (conn[j].url == NULL) break;
	}
	if (j == CONN_MAX) {
		fprintf(stderr, "Too many open connections\n");
		return NULL;
	}
	conn[j].url = sdb_strdup(strchr(url, ':')+1);
	sprintf(b, "%d", j);
	conn[j].index = sdb_strdup(b);
	conn[j].driver = i;
	conn[j].db = db;
	return conn[j].index;
}

int sdb_close(char *url)
{
	int i, j, n;

	if (strchr(url, ':') || !isdigit(*url)) return -1;

	j = atoi(url);
	if (j < 0 || j >= CONN_MAX) return -1;

	i = conn[j].driver;
	if (drivers[i].close && conn[j].db) n = (*drivers[i].close)(conn[j].db);
	else n = 0;

	sdb_free(conn[j].url);
	conn[j].url = NULL;
	sdb_free(conn[j].index);
	conn[j].index = NULL;
	return n;
}

static void sdb_exit(void)
{
	int i;
	char b[1024];

	for (i = 0; i < CONN_MAX; i++) {
		if (conn[i].url) {
			sprintf(b, "%d", i);
			sdb_close(b);
		}
	}
}

void sdb_init_sdbd(void)
{
	sdb_register_driver("sdbd", sdbd_driver, sdbd_open, sdbd_close);
}

void sdb_init(void)
{
	static int init_done = 0;

	if (init_done) return;
	init_done = 1;
	ndrivers = 0;
	sdb_init_sdbd();
	sdb_init_ingres();
        sdb_init_lago();
        sdb_init_mysql();
        sdb_init_sqlite();
        sdb_init_sqlite3();
        sdb_init_mimer();
        sdb_init_postgres();
        sdb_init_oracle();
	sdb_init_odbc();
	sdb_init_gdbm();
	sdb_init_msql();
	sdb_init_shsql();
	sdb_init_tds();
	sdb_init_text();
	atexit(sdb_exit);
}
